topics/icrc_X_multiple_canister_call.md

```diff
@@ -0,0 +1,331 @@
+            end
+        end
+
+        S ->> U: Showing signing message from consent message or blind message
+
+
+        alt Approved
+            U ->> S: Approve request
+            Note over S,C: ICRC X execution
+
+            loop until response length = request length
+                S ->> C: Submit canister call requests <br /> - if request has no 'waitFor', execute immediately in paralle <br /> - if request has 'waitFor', execute after dependencies finishes successfully
+                S ->> S: Wait for the canister call results
+
+
+                alt If tx opted to validate whether the request was successfully
+                    S ->> VC: Call check with validation canister using contentMap
+                    VC ->> S: Return success or failure as bool (new ICRC)
+                end
+
+                alt If the canister called is success
+                    S ->> S: Add call result to batch call response
+                else If the canister called is failed
+                    S ->> S: Add error to batch call response
+                    S ->> S: Add error with code `1001` for other non-execute request to batch call response
+                end
+            end
+
+            Note over S,C: End of the execution
+
+        S ->> U: Display success/failed message
+        S ->> RP: Return the batch call response
+
+        else Rejected
+            U ->> S: Reject request
+            S ->> U: Display reject message
+            S ->> RP: Error response: Action aborted (3001)
+        end
+    end
+```
+
+### Examples
+
+#### Successful Batch Call
+
+Approve request, swap and then call birdge method
+
+```json
+{
+  "id": 1,
+  "jsonrpc": "2.0",
+  "method": "icrcX_batch_call_canisters",
+  "params": {
+    "sender": "b7gqo-ulk5n-2kpo7-oalt7-p2kyl-o4j5l-kiuwo-eeybr-dab4l-ur6up-pqe",
+    "requests": [
+      {
+        "id": "1",
+        "canisterId": "eeddf-fqaaa-aaaao-a2hlq-ca",
+        "method": "icrc2_approve",
+        "arg": "RElETARte24AbAKzsNrDA2ithsqDBQFsA/vKAQKi3pTrBgHYo4yoDX0BAwEdV+ztKgq7E4l1ffuTuwEmw8AtYSjlrJ+WLO5ofQIAAMgB"
+      },
+      {
+        "id": "2",
+        "canisterId": "aaabb-fqaaa-aaaao-a2hlq-ca",
+        "method": "icrc2_approve",
+        "arg": "RElETARte24AbAKzsNrDA2ithsqDBQFsA/vKAQKi3pTrBgHYo4yoDX0BAwEdV+ztKgq7E4l1ffuTuwEmw8AtYSjlrJ+WLO5ofQIAAMgB"
+      },
+      {
+        "id": "3",
+        "canisterId": "xyzzz-fqaaa-aaaao-a2hlq-ca",
+        "method": "swap",
+        "arg": "RElETARte24AbAKzsNrDA2ithsqDBQFsA/vKAQKi3pTrBgHYo4yoDX0BAwEdV+ztKgq7E4l1ffuTuwEmw8AtYSjlrJ+WLO5ofQIAAMgB",
+        "waitFor": ["1", "2"]
+      },
+      ,
+      {
+        "id": "4",
+        "canisterId": "bbbbb-fqaaa-aaaao-a2hlq-ca",
+        "method": "bridge_to_eth",
+        "arg": "RElETARte24AbAKzsNrDA2ithsqDBQFsA/vKAQKi3pTrBgHYo4yoDX0BAwEdV+ztKgq7E4l1ffuTuwEmw8AtYSjlrJ+WLO5ofQIAAMgB",
+        "waitFor": ["3"],
+        "validateCanister": "xyzzz-fqaaa-aaaao-a2hlq-ca"
```

Comment:
Assuming that the validateCanister transform the response from request 4 into success true/false, it doesn't have any affect when defined in the last request since either way no other calls are executed after the last call.

So it should probably be moved to request 3 (maybe also 1 and 2?).

Is there a use case for having a different validateCanister for requests within the same batch? If not it might make more sense to move it up below the sender, this simplifies the signer implementation and enables it to validate multiple responses at once.

```json
{
  "id": 1,
  "jsonrpc": "2.0",
  "method": "icrcX_batch_call_canisters",
  "params": {
    "sender": "b7gqo-ulk5n-2kpo7-oalt7-p2kyl-o4j5l-kiuwo-eeybr-dab4l-ur6up-pqe",
    "validate": {
      canisterId: "xyzzz-fqaaa-aaaao-a2hlq-cai"
      method: "validate_batch_approve_swap_bridge"
    },
    "requests": [
      {
        "id": "1",
        "canisterId": "eeddf-fqaaa-aaaao-a2hlq-ca",
        "method": "icrc2_approve",
        "arg": "RElETARte24AbAKzsNrDA2ithsqDBQFsA/vKAQKi3pTrBgHYo4yoDX0BAwEdV+ztKgq7E4l1ffuTuwEmw8AtYSjlrJ+WLO5ofQIAAMgB"
      },
      {
        "id": "2",
        "canisterId": "aaabb-fqaaa-aaaao-a2hlq-ca",
        "method": "icrc2_approve",
        "arg": "RElETARte24AbAKzsNrDA2ithsqDBQFsA/vKAQKi3pTrBgHYo4yoDX0BAwEdV+ztKgq7E4l1ffuTuwEmw8AtYSjlrJ+WLO5ofQIAAMgB"
      },
      {
        "id": "3",
        "canisterId": "xyzzz-fqaaa-aaaao-a2hlq-ca",
        "method": "swap",
        "arg": "RElETARte24AbAKzsNrDA2ithsqDBQFsA/vKAQKi3pTrBgHYo4yoDX0BAwEdV+ztKgq7E4l1ffuTuwEmw8AtYSjlrJ+WLO5ofQIAAMgB",
        "waitFor": ["1", "2"]
      },
      ,
      {
        "id": "4",
        "canisterId": "bbbbb-fqaaa-aaaao-a2hlq-ca",
        "method": "bridge_to_eth",
        "arg": "RElETARte24AbAKzsNrDA2ithsqDBQFsA/vKAQKi3pTrBgHYo4yoDX0BAwEdV+ztKgq7E4l1ffuTuwEmw8AtYSjlrJ+WLO5ofQIAAMgB",
        "waitFor": ["3"]
      }
    ]
  }
}
```

The `validate_batch_approve_swap_bridge` method of canister `xyzzz-fqaaa-aaaao-a2hlq-cai` would be called twice, first with an array containing the responses from request 1 and 2 and later with an array containing the response from request 3. Keeping latency to a minimum by limiting the number of validation calls to the number of unique `waitFor` combinations.

Reply:
It's like ICRC21 for response registry canister. Yeah ok, I will talk this with Mason

Reply:
Is there any flag to not making a new call to validate the response?

If I'm a developer making the swap and bridge, I should do the validation for allowance before executing the swap and bridge. 

we can put `ignore` 
```
 "validate": {
      canisterId: "xyzzz-fqaaa-aaaao-a2hlq-cai"
      method: "validate_batch_approve_swap_bridge"
      ignore: ["3", "4"]
   },
```

or `validate` is optional here

Reply:
I would argue that validating the last request is never needed, since there is no request after the last request that depends on its outcome.

In case you don't want to validate 3, then there's no reason to make 4 wait for 3 in the first place. You could argue that you'd want to make 4 wait for 3 independent of the outcome of 3 but that is an edge case I wouldn't cover here, instead simply always validate the responses other calls have waited for. Else we end up with more than two behaviors (parallel or sequential based on outcome) which makes things more complicated than needed.

As for making validate optional overall, I think that makes sense but then we'd need to decide what happens if there are waitFor properties defined, either:
- Throw error "validate is required when call depends on each other"
- Expect wallet to fallback to either defining success based on ICRC knowledge or "the call responded at least".

Considering that validation can be implemented without any reliance on devs that implemented the target canisters or wallets implementing ICRC knowledge correctly. I would argue to go for the first option, it guarantees consistent behavior across all signers since the dapp developer itself is in control.